# Getting started

### CLEF 2025 - CheckThat! Lab  - Task 4 Scientific Web Discourse - Subtask 4b (Scientific Claim Source Retrieval)



- **https://www.nltk.org/**
- **https://spacy.io/docs**

In [1]:
pip install nltk spacy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
!pip install spacy
!python -m spacy download en_core_web_sm




[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     --------------------------------------- 0.0/12.8 MB 435.7 kB/s eta 0:00:30
     --------------------------------------- 0.1/12.8 MB 544.7 kB/s eta 0:00:24
     --------------------------------------- 0.1/12.8 MB 654.9 kB/s eta 0:00:20
     --------------------------------------- 0.2/12.8 MB 701.4 kB/s eta 0:00:19
      -------------------------------------- 0.2/12.8 MB 692.9 kB/s eta 0:00:19
      -------------------------------------- 0.2/12.8 MB 722.1 kB/s eta 0:00:18
      -------------------------------------- 0.3/12.8 MB 744.2 kB/s eta 0:00:17
      -------------------------------------- 0.3/12.8 MB 756.6 kB/s eta 0:00:17
     - ------------------------------------- 0.4/12.8 MB 796.7 kB/s eta 0:00:16
     - ------------------------------------- 0.4/12.8 MB 798.7 kB/s eta 0:00:16
     - ------------------------------------- 0.4/12.8 MB


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import numpy as np
import pandas as pd


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.4 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykern

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.4 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\HP-ZBOOK i7\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykern

AttributeError: _ARRAY_API not found

## 1.a) Import the collection set
The collection set contains metadata of CORD-19 academic papers.

The preprocessed and filtered CORD-19 dataset is available on the Gitlab repository here: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b

Participants should first download the file then upload it on the Google Colab session with the following steps.


In [4]:
# 1) Download the collection set from the Gitlab repository: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b
# 2) Drag and drop the downloaded file to the "Files" section (left vertical menu on Colab)
# 3) Modify the path to your local file path
PATH_COLLECTION_DATA = '../subtask_4b/subtask4b_collection_data.pkl'

In [5]:
df_collection = pd.read_pickle(PATH_COLLECTION_DATA)

### Dataframe Information (`df_collection.info()`):

The dataframe `df_collection` contains **7,718 entries** (rows) and **17 columns**. This is the metadata for **7,718 papers** in the CORD-19 dataset.


In [6]:
df_collection.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7718 entries, 162 to 1056448
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   cord_uid          7718 non-null   object        
 1   source_x          7718 non-null   object        
 2   title             7718 non-null   object        
 3   doi               7677 non-null   object        
 4   pmcid             4959 non-null   object        
 5   pubmed_id         6233 non-null   object        
 6   license           7718 non-null   object        
 7   abstract          7718 non-null   object        
 8   publish_time      7715 non-null   object        
 9   authors           7674 non-null   object        
 10  journal           6668 non-null   object        
 11  mag_id            0 non-null      float64       
 12  who_covidence_id  528 non-null    object        
 13  arxiv_id          20 non-null     object        
 14  label             7718 n

In [7]:
df_collection.head()

cord_uid source_x                                              title  \
162   umvrwgaw      PMC  Professional and Home-Made Face Masks Reduce E...   
611   spiud6ok      PMC                               The Failure of R (0)   
918   aclzp3iy      PMC  Pulmonary sequelae in a patient recovered from...   
993   ycxyn2a2      PMC  What was the primary mode of smallpox transmis...   
1053  zxe95qy9      PMC  Lessons from the History of Quarantine, from P...   

                               doi       pmcid pubmed_id      license  \
162   10.1371/journal.pone.0002618  PMC2440799  18612429        cc-by   
611            10.1155/2011/527610  PMC3157160  21860658        cc-by   
918        10.4103/0970-2113.99118  PMC3424870  22919170  cc-by-nc-sa   
993       10.3389/fcimb.2012.00150  PMC3509329  23226686        cc-by   
1053        10.3201/eid1902.120312  PMC3559034  23343512        no-cc   

                                               abstract publish_time  \
162   BACKGROUND: Governments are preparing for a po...   2008-07-09   
611   The basic reproductive ratio, R (0), is one of...   2011-08-16   
918   The pandemic of swine flu (H1N1) influenza spr...         2012   
993   The mode of infection transmission has profoun...   2012-11-29   
1053  In the new millennium, the centuries-old strat...   2013-02-03   

                                                authors  \
162   van der Sande, Marianne; Teunis, Peter; Sabel,...   
611       Li, Jing; Blakeley, Daniel; Smith?, Robert J.   
918   Singh, Virendra; Sharma, Bharat Bhushan; Patel...   
993                                   Milton, Donald K.   
1053                                  Tognotti, Eugenia   

                          journal  mag_id who_covidence_id arxiv_id     label  \
162                      PLoS One     NaN              NaN      NaN  umvrwgaw   
611       Comput Math Methods Med     NaN              NaN      NaN  spiud6ok   
918                    Lung India     NaN              NaN      NaN  aclzp3iy   
993   Front Cell Infect Microbiol     NaN              NaN      NaN  ycxyn2a2   
1053             Emerg Infect Dis     NaN              NaN      NaN  zxe95qy9   

           time       timet  
162  2008-07-09  1215561600  
611  2011-08-16  1313452800  
918  2012-01-01  1325376000  
993  2012-11-29  1354147200  
1053 2013-02-03  1359849600

In this process, I performed several text preprocessing steps on the title and abstract columns of the dataset df_collection** First, I removed any URLs using a regular expression. I then removed punctuation and digits, converted the text to lowercase, and eliminated common stopwords, except for those found in the important_terms list, which includes relevant medical terms like disease names . To further lemmatization the text, I applied stemming to all words, except for those in the important_terms list. Additionally, I used the **spaCy** library to extract named entities from both the title and abstract columns. I also cleaned up extra spaces by reducing multiple spaces to a single one . As a result of these preprocessing steps, I created new columns for the cleaned title** and abstract, as well as for the extracted entities from both columns.

In [ ]:
import re
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import spacy

stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()
nlp = spacy.load("en_core_web_sm")


important_terms = {
    'covid', 'covid-19', 'hiv', 'rna', 'sars', 'r0', 'h1n1', 'who', 'cdc', 
    'micro', 'kg', 'm/s', 'flu', 'cancer', 'aids', 'diabetes', 'malaria', 
    'tuberculosis', 'hepatitis', 'pneumonia', 'leprosy', 'arthritis', 'asthma', 
    'hypertension', 'obesity', 'influenza', 'hiv', 'hepatitis-b', 'coronavirus', 
    'outbreak', 'pandemic', 'endemic', 'vaccine', 'antiviral', 'antibiotic', 
    'surgical', 'gene', 'genome', 'mutation', 'pathogen'
}



def preprocess_text(text):
    
    
    text = re.sub(r'http\S+|www\S+', '', text)
    
    
     
    text = ' '.join([word if word in important_terms else re.sub(r'\d+', '', word) for word in text.split()])

     
    text = re.sub(r'\s+', ' ', text).strip()


    text = text.translate(str.maketrans('', '', string.punctuation))

    
 
    text = text.lower()
    
    text = ' '.join([word for word in text.split() if word not in stop_words or word in important_terms])
    
    text = ' '.join([stemmer.stem(word) if word not in important_terms else word for word in text.split()])
    
    
    
    doc = nlp(text)
    entities = [ent.text for ent in doc.ents] 
    

    return text, entities


df_collection['cleaned_title'], df_collection['entities_in_title'] = zip(*df_collection['title'].apply(preprocess_text))
df_collection['cleaned_abstract'], df_collection['entities_in_abstract'] = zip(*df_collection['abstract'].apply(preprocess_text))






                                                  title  \
162   Professional and Home-Made Face Masks Reduce E...   
611                                The Failure of R (0)   
918   Pulmonary sequelae in a patient recovered from...   
993   What was the primary mode of smallpox transmis...   
1053  Lessons from the History of Quarantine, from P...   

                                          cleaned_title  \
162   profession homemad face mask reduc exposur res...   
611                                            failur r   
918           pulmonari sequela patient recov swine flu   
993    primari mode smallpox transmiss implic biodefens   
1053           lesson histori quarantin plagu influenza   

                                               abstract  \
162   BACKGROUND: Governments are preparing for a po...   
611   The basic reproductive ratio, R (0), is one of...   
918   The pandemic of swine flu (H1N1) influenza spr...   
993   The mode of infection transmission has profoun..

In this part of the code, I created a function extract_entities that uses the spaCy library to perform Named Entity Recognition (NER) on the input text. The function processes the text and extracts the recognized entities as a list. I then applied this function to the title and abstract columns of the df_collection dataset, storing the extracted entities in two new columns: entities_in_title and entities_in_abstract. These columns now contain lists of entities that were identified in the  title and `abstract of each record.


In [10]:
def extract_entities(text):
    doc = nlp(text)
    entities = [ent.text for ent in doc.ents]
    return entities

df_collection['entities_in_title'] = df_collection['title'].apply(extract_entities)
df_collection['entities_in_abstract'] = df_collection['abstract'].apply(extract_entities)


In [26]:

filtered_df = df_collection[['title', 'abstract', 'cleaned_title', 'cleaned_abstract', 'entities_in_title', 'entities_in_abstract']]



In [30]:
filtered_df.head(10)

title  \
162   Professional and Home-Made Face Masks Reduce E...   
611                                The Failure of R (0)   
918   Pulmonary sequelae in a patient recovered from...   
993   What was the primary mode of smallpox transmis...   
1053  Lessons from the History of Quarantine, from P...   
1589  Anxiety and Depression: Linkages with Viral Di...   
2069  Effects of Ultraviolet Germicidal Irradiation ...   
2843  Secretome of Intestinal Bacilli: A Natural Gua...   
2952  Long term outcomes in survivors of epidemic In...   
3044  Far-UVC light: A new tool to control the sprea...   

                                               abstract  \
162   BACKGROUND: Governments are preparing for a po...   
611   The basic reproductive ratio, R (0), is one of...   
918   The pandemic of swine flu (H1N1) influenza spr...   
993   The mode of infection transmission has profoun...   
1053  In the new millennium, the centuries-old strat...   
1589  Anxiety and mood disorders are common in the g...   
2069  The ability to disinfect and reuse disposable ...   
2843  Current studies of human gut microbiome usuall...   
2952  Patients who survive influenza A (H7N9) virus ...   
3044  Airborne-mediated microbial diseases such as i...   

                                          cleaned_title  \
162   profession homemad face mask reduc exposur res...   
611                                            failur r   
918           pulmonari sequela patient recov swine flu   
993    primari mode smallpox transmiss implic biodefens   
1053           lesson histori quarantin plagu influenza   
1589                anxieti depress linkag viral diseas   
2069  effect ultraviolet germicid irradi uvgi n resp...   
2843     secretom intestin bacilli natur guard patholog   
2952  long term outcom survivor epidem influenza hn ...   
3044  faruvc light new tool control spread airbornem...   

                                       cleaned_abstract  \
162   background govern prepar potenti influenza pan...   
611   basic reproduct ratio r one fundament concept ...   
918   pandemic swine flu hn influenza spread involv ...   
993   mode infect transmiss profound implic effect c...   
1053  new millennium centuriesold strategi quarantin...   
1589  anxieti mood disord common gener popul countri...   
2069  abil disinfect reus dispos n filter facepiec r...   
2843  current studi human gut microbiom usual consid...   
2952  patient who surviv influenza hn viru infect ri...   
3044  airbornemedi microbi diseas influenza tubercul...   

                                      entities_in_title  \
162   [Professional and Home-Made Face Masks Reduce ...   
611                                                  []   
918                                         [Pulmonary]   
993                                                  []   
1053                [the History of Quarantine, Plague]   
1589                                   [Viral Diseases]   
2069      [UVGI, N95 Respirator Filtration Performance]   
2843                                      [Pathologies]   
2952                                                 []   
3044                                                 []   

                                   entities_in_abstract  
162                                [PRINCIPAL, Outward]  
611                                     [1, 1, 1, 1, 1]  
918                                          [3 months]  
993          [a few seconds, the last decades, Variola]  
1053  [the new millennium, centuries-old, 2003, just...  
1589  [the United States, Great Britain, herpes simp...  
2069  [N95, UVGI, one, UVGI, four, N95, UVGI, 120–95...  
2843            [GIT, two, Bacilli, colonize, GIT, GIT]  
2952  [Fifty-six, 2-year, 6 months, the first six mo...  
3044                      [207–222, first, 2, 222, 95%]

## 1.b) Import the query set

The query set contains tweets with implicit references to academic papers from the collection set.

The preprocessed query set is available on the Gitlab repository here: https://gitlab.com/checkthat_lab/clef2025-checkthat-lab/-/tree/main/task4/subtask_4b

Participants should first download the file then upload it on the Google Colab session with the following steps.

In [31]:

PATH_QUERY_TRAIN_DATA = '../subtask_4b/subtask4b_query_tweets_train.tsv'
PATH_QUERY_DEV_DATA = '../subtask_4b/subtask4b_query_tweets_dev.tsv'

In [32]:
df_query_train = pd.read_csv(PATH_QUERY_TRAIN_DATA, sep = '\t')
df_query_dev = pd.read_csv(PATH_QUERY_DEV_DATA, sep = '\t')

**tweettext** column was cleaned using the **preprocess_text** function 

In [33]:

df_query_train['cleaned_tweet_text'] = df_query_train['tweet_text'].apply(preprocess_text)
df_query_train['cleaned_tweet_text'] = df_query_train['cleaned_tweet_text'].fillna('').astype(str)


df_query_dev['cleaned_tweet_text'] = df_query_dev['tweet_text'].apply(preprocess_text)
df_query_dev['cleaned_tweet_text'] = df_query_dev['cleaned_tweet_text'].fillna('').astype(str)


In [34]:

print(df_query_dev.columns.tolist())


['post_id', 'tweet_text', 'cord_uid', 'cleaned_tweet_text']


In [35]:
df_query_train.head()

post_id                                         tweet_text  cord_uid  \
0        0  Oral care in rehabilitation medicine: oral vul...  htlvpvz5   
1        1  this study isn't receiving sufficient attentio...  4kfl29ul   
2        2  thanks, xi jinping. a reminder that this study...  jtwb17u8   
3        3  Taiwan - a population of 23 million has had ju...  0w9k8iy1   
4        4  Obtaining a diagnosis of autism in lower incom...  tiqksd69   

                                  cleaned_tweet_text  
0  ('oral care rehabilit medicin oral vulner oral...  
1  ('studi isnt receiv suffici attent reveal blac...  
2  ('thank xi jinp remind studi conclud nonpharma...  
3  ('taiwan popul million case amp death widespre...  
4  ('obtain diagnosi autism lower incom countri t...

We apply the **extract_entities** function to the **tweet_text** column and store the resulting entities in a single column for both the training and development datasets. 

In [36]:
df_query_train['entities_in_tweet_text_train'] = df_query_train['tweet_text'].apply(extract_entities)
df_query_train['entities_in_tweet_text_train'] = df_query_train['entities_in_tweet_text_train'].fillna('').astype(str)


df_query_dev['entities_in_tweet_text_dev'] = df_query_dev['tweet_text'].apply(extract_entities)
df_query_dev['entities_in_tweet_text_dev'] = df_query_dev['entities_in_tweet_text_dev'].fillna('').astype(str)


# 2)The following code runs a BM25 after preprcoessing 



In [37]:
!pip install rank_bm25
from rank_bm25 import BM25Okapi



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [38]:

corpus = df_collection[['cleaned_title', 'cleaned_abstract', 'entities_in_title', 'entities_in_abstract']].apply(
    lambda x: f"{x['cleaned_title']} {x['cleaned_abstract']} {' '.join(x['entities_in_title'])} {' '.join(x['entities_in_abstract'])}",
    axis=1
).tolist()

cord_uids = df_collection[:]['cord_uid'].tolist()

tokenized_corpus = [doc.split(' ') for doc in corpus]

bm25 = BM25Okapi(tokenized_corpus)


In [39]:
def get_top_cord_uids(query):
  text2bm25top = {}
  if query in text2bm25top.keys():
      return text2bm25top[query]
  else:
      tokenized_query = query.split(' ')
    

      #based on the tokenized query we call the BM25 to compute the relevance score for each document relevant to the query 
      doc_scores = bm25.get_scores(tokenized_query)
      # it sort the doc scores in descending order 
      indices = np.argsort(-doc_scores)[:5]
      #This line uses the indices from the previous step to retrieve the actual document IDs 
      bm25_topk = [cord_uids[x] for x in indices]
      

      text2bm25top[query] = bm25_topk
      return bm25_topk



In [40]:


df_query_train['final_query'] = df_query_train['cleaned_tweet_text'] + ' ' + df_query_train['entities_in_tweet_text_train']
df_query_dev['final_query'] = df_query_dev['cleaned_tweet_text'] + ' ' + df_query_dev['entities_in_tweet_text_dev']


df_query_train['bm25_topk'] = df_query_train['final_query'].apply(lambda x: get_top_cord_uids(x))
df_query_dev['bm25_topk'] = df_query_dev['final_query'].apply(lambda x: get_top_cord_uids(x))





# 3) Evaluating the baseline
The following code evaluates the BM25 retrieval baseline on the query set using the Mean Reciprocal Rank score (MRR@5).

In [41]:
# Evaluate retrieved candidates using MRR@k
def get_performance_mrr(data, col_gold, col_pred, list_k = [1, 5, 10]):
    d_performance = {}
    for k in list_k:
        data["in_topx"] = data.apply(lambda x: (1/([i for i in x[col_pred][:k]].index(x[col_gold]) + 1) if x[col_gold] in [i for i in x[col_pred][:k]] else 0), axis=1)
        #performances.append(data["in_topx"].mean())
        d_performance[k] = data["in_topx"].mean()
    return d_performance


In [42]:
# Evaluate retrieved candidates using MRR@k
results_train = get_performance_mrr(df_query_train, 'cord_uid', 'bm25_topk')
results_dev = get_performance_mrr(df_query_dev, 'cord_uid', 'bm25_topk')

# Printed MRR@k results
print(f"Results on the train set: {results_train}")
print(f"Results on the dev set: {results_dev}")


Results on the train set: {1: np.float64(0.5221349101377111), 5: np.float64(0.5746868435384734), 10: np.float64(0.5746868435384734)}
Results on the dev set: {1: np.float64(0.5185714285714286), 5: np.float64(0.5740119047619048), 10: np.float64(0.5740119047619048)}


| Set   | Top-K | Score     |
|--------|--------|-----------|
| Train | 1     | 0.5221 |
| Train | 5     | 0.5747 |
| Train | 10    | 0.5747 |
| Dev   | 1     | 0.5186 |
| Dev   | 5     | 0.5740 |
| Dev   | 10    | 0.5740 |


# 4) Exporting results to prepare the submission on Codalab

In [43]:
df_query_dev['preds'] = df_query_dev['bm25_topk'].apply(lambda x: x[:5])

In [44]:
df_query_dev[['post_id', 'preds']].to_csv('predictions.tsv', index=None, sep='\t')